File: 2022/rust/src/bin/day05.rs
use aoc2022::Solve;
use itertools::Itertools;
use std::collections::VecDeque;
use std::io::{BufRead, BufReader, Read};

struct Solution {
    stacks: Vec<Vec<char>>,
    procedure: Vec<(usize, usize, usize)>,
}

impl Solution {
    fn top_crates(&self, f: impl Fn(&mut VecDeque<char>) -> Option<char>) -> String {
        let mut stacks = self.stacks.clone();
        for &(count, from, to) in &self.procedure {
            let mut vd = (0..count)
                .filter_map(|_| stacks[from - 1].pop())
                .collect::<VecDeque<_>>();
            while let Some(c) = f(&mut vd) {
                stacks[to - 1].push(c);
            }
        }
        stacks.iter().filter_map(|s| s.last()).collect()
    }
}

impl Solve for Solution {
    type Answer1 = String;
    type Answer2 = String;

    fn new(r: impl Read) -> Self {
        let lines = BufReader::new(r)
            .lines()
            .filter_map(Result::ok)
            .collect::<Vec<_>>();
        let parts = lines.split(String::is_empty).collect::<Vec<_>>();
        let mut stacks = vec![Vec::new(); (parts[0][0].len() + 1) / 4];
        for line in parts[0].iter().rev().skip(1) {
            for (i, c) in line.chars().skip(1).step_by(4).enumerate() {
                if c.is_ascii_uppercase() {
                    stacks[i].push(c);
                }
            }
        }
        Self {
            stacks,
            procedure: parts[1]
                .iter()
                .filter_map(|s| {
                    s.split_ascii_whitespace()
                        .skip(1)
                        .step_by(2)
                        .filter_map(|s| s.parse().ok())
                        .collect_tuple()
                })
                .collect(),
        }
    }
    fn part1(&self) -> Self::Answer1 {
        self.top_crates(VecDeque::pop_front)
    }
    fn part2(&self) -> Self::Answer2 {
        self.top_crates(VecDeque::pop_back)
    }
}

fn main() {
    let solution = Solution::new(std::io::stdin().lock());
    println!("Part 1: {}", solution.part1());
    println!("Part 2: {}", solution.part2());
}

#[cfg(test)]
mod tests {
    use super::*;

    fn example_input() -> &'static [u8] {
        r"
    [D]    
[N] [C]    
[Z] [M] [P]
 1   2   3 

move 1 from 2 to 1
move 3 from 1 to 3
move 2 from 2 to 1
move 1 from 1 to 2
"[1..]
            .as_bytes()
    }

    #[test]
    fn example1() {
        assert_eq!("CMZ", Solution::new(example_input()).part1());
    }

    #[test]
    fn example2() {
        assert_eq!("MCD", Solution::new(example_input()).part2());
    }
}

File: 2022/ocaml/lib/day05.ml
open Base
open Solution

module Solution : Solve = struct
  type t = (char list -> char list) -> string

  let parse input =
    let hd, tl =
      Stdio.In_channel.input_lines input
      |> List.split_while ~f:(Fn.non String.is_empty)
    in
    let stacks =
      let stack_of i =
        let extract_crate s =
          String.get s ((i * 4) + 1) |> function
          | 'A' .. 'Z' as c -> Some c
          | _ -> None
        in
        List.filter_map hd ~f:extract_crate
      in
      let len = ((List.last_exn hd |> String.length) + 1) / 4 in
      List.range 0 len |> List.map ~f:stack_of
    in
    let procedure =
      let parse_procedure s =
        Caml.Scanf.sscanf s "move %d from %d to %d" (fun num from_idx to_idx ->
            (num, from_idx - 1, to_idx - 1))
      in
      List.drop tl 1 |> List.map ~f:parse_procedure
    in
    fun f ->
      let move stacks (num, src, dst) =
        let hd, tl = List.split_n stacks.(src) num in
        stacks.(dst) <- f hd @ stacks.(dst);
        stacks.(src) <- tl;
        stacks
      in
      List.fold procedure ~init:(Array.of_list stacks) ~f:move
      |> Array.filter_map ~f:List.hd
      |> Array.to_list |> String.of_char_list

  let part1 top_crates = top_crates List.rev |> answer_of_string
  let part2 top_crates = top_crates Fn.id |> answer_of_string
end
